[FLINK-32137] Added filtering of lambdas when building the flame graph

netvl opened the pull request:
## What is the purpose of the change

Because lambda "class names" can be different across different JVMs, the flame graph in Flink UI is often unreadable, especially with high parallelism with large number of task managers (there are 5 task managers here and therefore 5 "columns", but if the application has 50+ task managers, this becomes completely unreadable):
<img width="1337" alt="bad-1" src="https://github.com/apache/flink/assets/280456/a45ec88d-e5d6-4905-93e3-7d20f872481c">

This can happen in both Flink's internal logic which uses lambdas:
<img width="1337" alt="bad-3" src="https://github.com/apache/flink/assets/280456/b1f6ca32-989f-4715-bf5e-20c2ae782524">

and in user code:
<img width="1337" alt="bad-2" src="https://github.com/apache/flink/assets/280456/765e528a-47d1-4796-8b06-96ea7be59f3d">

The subtask view does help partially, but it is bad for understanding of the big picture, when you want to check what happens within the task as a whole.

This PR adds filtering logic to the code which collects stack traces which are subsequently used to build flame graphs. This logic modifies the stack trace elements which correspond to lambda invocations by replacing unique parts with static values, so in the final stack trace used for flame graphs these lambda calls are the same across different JVMS:
![good](https://github.com/apache/flink/assets/280456/f80ce6ae-6d3f-4f76-a75a-18e9bffd8fdf)

## Brief change log

- Added modification of stack traces to fix lambda-related entries
- Added tests for this functionality

## Verifying this change

This change added tests and can be verified as follows:

  - Manually verified the change by running a 6 node cluster with 1 JobManager and 5 TaskManagers, and captured flame graphs of a simple job with a parallel cpu-heavy map operation before and after the change. The pictures are available above.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/messages/ThreadInfoSample.java:
```diff
@@ -38,7 +42,7 @@ public class ThreadInfoSample implements Serializable {
     private ThreadInfoSample(Thread.State threadState, StackTraceElement[] stackTrace) {
         this.threadState = threadState;
-        this.stackTrace = stackTrace;
+        this.stackTrace = cleanLambdaNames(stackTrace);
```
The `ThreadInfoSample` is common class, should we clean lambda for all ThreadInfoSamples by default? or just clean it for flame graph?

netvl replied:
I looked into how this class is used, and it appears that its stacktrace-related functionality is only used for flame graphs (and I think it was added for flame graphs specifically).
Doing this cleanup as close to the place where stack traces are created results in cleaner code, and I personally think it is highly unlikely for the class name change to affect anything.

1996fanrui replied:
Currently, the `ThreadInfoSample ` is just used for flame graph. However, it's a common class.

The lambda name just effects the flame graph, so I prefer clean LambdaNames at the flame graph side, it's more clear, WDYT?

netvl replied:
Fair enough. I'll check where it can be done in the flame graph logic.

1996fanrui reviewed (commented):
Hi @netvl , thanks for the contribution!

Overall, I think this fix works well. However, I'm not sure whether removing the lambda makes sense. Because I found the method of lambda is useful. Here is an example, the call trace is :

- Task.runWithSystemExitMonitoring()
- action.run() ------ The action is a lambda
- StreamTask.invoke

If we remove the action.run(), the flame graph is `Task.runWithSystemExitMonitoring()` calls `StreamTask.invoke` directly. However, we cannot find the `invoke` inside of  `runWithSystemExitMonitoring`. The flamegraph will be weird.

Looking forward to your feedback, thanks~

<img width="1008" alt="image" src="https://github.com/apache/flink/assets/38427477/adbcbedf-31ef-4da4-b7cb-4faa41ab3882">


<img width="595" alt="image" src="https://github.com/apache/flink/assets/38427477/302ffc38-164b-434d-b90b-d3861c2d6cfe">

1996fanrui commented:
The CI fails due to checkstyle, please run the `mvn spotless:apply` to fix it.

https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=50356&view=logs&j=52b61abe-a3cc-5bde-cc35-1bbe89bb7df5&t=54421a62-0c80-5aad-3319-094ff69180bb&l=2929

netvl commented:
@1996fanrui I actually discovered the spotless issue locally but forgot to commit it >_<

> Because I found the method of lambda is useful

My initial implementation was actually replacing lambda names like this:
```
org.apache.flink.streaming.runtime.io.RecordProcessorUtils$$Lambda$773/0x00000001007f84a0
-->
org.apache.flink.streaming.runtime.io.RecordProcessorUtils$$Lambda$0/0x0000000000000000
```

However, the code was relatively complex so I decided to go with removal, which was much simpler. But I did miss that the method name is there too, and it can indeed give some useful context. I can switch the implementation back to "name unification", I guess.

netvl commented:
Sorry for the delay, I was attending a conference last week and wasn't able to finish up my contribution.

When bringing the cleanup logic closer to where the flame graphs are build makes this logic trickier to test. 
I will try to finalize it next week.

netvl commented:
@1996fanrui I've pushed an update which moves the stack frame filtering logic to the place where the flame graphs are built, and also switched to "fix" the class names instead of outright remove the trace entries. Please take a look)

1996fanrui reviewed (commented):
Hi @netvl , thanks for the update.

LGTM. Could you help paste a picture here?

1996fanrui commented on flink-runtime/src/test/java/org/apache/flink/runtime/webmonitor/threadinfo/VertexFlameGraphFactoryTest.java:
```diff
@@ -0,0 +1,80 @@
+package org.apache.flink.runtime.webmonitor.threadinfo;
```
This file doesn't includes the apache license.

netvl replied:
Fixed, thanks!

1996fanrui reviewed (commented):
Could you add [flamegraph] as the module name to the commit message?

netvl commented:
@1996fanrui added, please check that I did it correctly) not sure what do you mean by pasting a picture though.

1996fanrui commented:
> @1996fanrui added, please check that I did it correctly) not sure what do you mean by pasting a picture though.

Hi @netvl , all pictures that you provided are removed the lambda. The final strategy is convert the lambda name instead of removing it. So could you provide a picture based on current PR? It should keep these lambda lines and just rename the lambda.

1996fanrui reviewed (changes requested):
https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=51393&view=logs&j=52b61abe-a3cc-5bde-cc35-1bbe89bb7df5&t=54421a62-0c80-5aad-3319-094ff69180bb&l=2705

CI still fails due to the checkstyle

netvl commented:
@1996fanrui sorry for the delay! I fixed the checkstyle issues (`mvn checkstyle:check` now passes locally), and I fixed the picture and the description of the PR. Please review, thanks!)

1996fanrui commented:
@flinkbot run azure

1996fanrui commented:
> The CI job appears to have failed - https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=51797&view=results But I don't understand how to see the exact cause of it (looks like some error from bash?)

Could you rebase the apache/flink/master branch?

BTW, the Flink-1.18 has been code freezed, and this PR is an improvement, would you mind merging it at Flink-1.19?

1996fanrui reviewed (approved):
Thanks @netvl 's contribution and update.

LGTM, I will merge it after flink checkout to 1.19-SNAPSHOT.

netvl commented:
Thank you very much!
